pkg/query/assetsapi.go

```diff
@@ -0,0 +1,104 @@
+package query
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/kubecost/cost-model/pkg/kubecost"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+type assetResponse struct {
+	Code int                 `json:"code"`
+	Data []map[string]Assets `json:"data"`
+}
+
+type AssetParameters struct {
+	RestConfig *rest.Config
+	Ctx        context.Context
+
+	KubecostNamespace  string
+	ServiceName        string
+	Window             string
+	Aggregate          string
+	DisableAdjustments bool
+	Accumulate         bool
+	UseProxy           bool
+	FilterTypes        string
+}
+
+// QueryAssets queries /model/assets by proxying a request to Kubecost
+// through the Kubernetes API server if useProxy is true or, if it isn't, by
+// temporarily port forwarding to a Kubecost pod.
+func QueryAssets(p AssetParameters) ([]map[string]Assets, error) {
+
+	// aggregate, accumulate, and disableAdjustments are hardcoded;
+	// as other asset types are added in to be filtered by, this may change,
+	// but for now anything beyond isn't needed.
+
+	requestParams := map[string]string{
+		"window":             p.Window,
+		"aggregate":          "",
+		"accumulate":         "true",
+		"disableAdjustments": "true",
+		"filterTypes":        p.FilterTypes,
+	}
+
+	if p.Aggregate != "" {
+		requestParams["aggregate"] = p.Aggregate
+	}
+
+	var bytes []byte
+	var err error
+	if p.UseProxy {
+		clientset, err := kubernetes.NewForConfig(p.RestConfig)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create clientset: %s", err)
+		}
+
+		bytes, err = clientset.CoreV1().Services(p.KubecostNamespace).ProxyGet("", p.ServiceName, "9090", "/model/assets", requestParams).DoRaw(p.Ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to proxy get kubecost. err: %s; data: %s", err, bytes)
+		}
+	} else {
+		bytes, err = portForwardedQueryService(p.RestConfig, p.KubecostNamespace, p.ServiceName, "model/assets", requestParams, p.Ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to port forward query: %s", err)
+		}
+	}
+
+	var ar assetResponse
+	err = json.Unmarshal(bytes, &ar)
+	if err != nil {
+		return ar.Data, fmt.Errorf("failed to unmarshal allocation response: %s", err)
+	}
+
+	return ar.Data, nil
+}
+
+type Assets struct {
```

Comment:
I'd recommend you call this `type AssetNode` or `type Node` instead, as not all assets have these fields. When we add more asset types, we can use something like struct embedding for the common parameters, but we'll leave that alone for now.

Reply:
Which means you'll have to make the output table function just take a `Node`. That'll be a problem later, but let's cross that bridge when we come to it.